File: server/session_handler/api_session/stream_layer.py
from dataclasses import dataclass
import os
from queue import Queue, Empty
from threading import Event, Lock, Thread

from core import session_settings
from session_handler.data_channels import UDPChannel


@dataclass(frozen=True)
class RequestFrame:
    size: int


class __WorkerStreamLayer:
    def __process_request(self, request_frame: RequestFrame):
        """
        process a single request of stream bytes from client

        args:
            request_frame: RequestFrame -> a frame received from queue to be processed
        """
        with self._lock:
            data = self._video_file.read(request_frame.size)
            self._udp_channel.send_data(data)

    def __handler_video(self):
        while True:
            with self._pause_lock:
                pass
            if self._kill_event.is_set():
                return
            try:
                request_frame = self._requests_queue.get(block=False)
            except Empty:
                continue
            self.__process_request(request_frame)

    def __init__(self, udp_channel: UDPChannel, video_path: str) -> None:
        self._udp_channel = udp_channel
        self._video_path = video_path
        self._video_file = open(video_path, "rb")
        self._file_size = os.path.getsize(video_path)
        self._requests_queue: Queue[RequestFrame] = Queue(session_settings.max_requests)
        self._lock = Lock()
        self._pause_lock = Lock()
        self._kill_event = Event()
        self._worker_thread = Thread(target=self.__handler_video)
        self._worker_thread.start()


class StreamLayer(__WorkerStreamLayer):
    def add_request(self, request_frame: RequestFrame):
        self._requests_queue.put(request_frame, block=True)

    def update_offset(self, offset: int) -> bool:
        if offset >= self._file_size:
            return False
        with self._lock:
            self._video_file.seek(offset)
            return True
    
    def seek_forward_offset(self):
        current_offset = self._video_file.tell()
        new_offset = current_offset + int(self._file_size * 0.01)
        
        if new_offset >= self._file_size:
            print("Error: Cannot seek beyond the end of the video.")
            return False
        
        return self.update_offset(new_offset)

    def seek_backward_offset(self):
        current_offset = self._video_file.tell()
        new_offset = max(0, current_offset - int(self._file_size * 0.01))
        
        if new_offset < 0:
            new_offset = 0  # start video from the beginning if offset is less than 0

        return self.update_offset(new_offset)

    def pause(self):
        if self._pause_lock.locked():
            self._pause_lock.release()
            return
        self._pause_lock.acquire()

    def stop(self):
        with self._lock:
            self._video_file.close()
        self._kill_event.set()
        self._worker_thread.join()
        self._udp_channel.close()

    def __del__(self):
        self.stop()

File: serverjava/src/session_handler/apiSession/StreamLayer.java
package session_handler.apiSession;

import java.io.*;
import java.util.concurrent.*;
import java.util.concurrent.locks.*;

import session_handler.Data_channel.UDPChannel;
import session_handler.core.Settings.SessionSettings;

public class StreamLayer {

    // Define the request frame class
    public static class RequestFrame {
        private final int size;

        public RequestFrame(int size) {
            this.size = size;
        }

        public int getSize() {
            return size;
        }
    }

    // WorkerStreamLayer as a private static inner class
    private static class WorkerStreamLayer {

        private final UDPChannel udpChannel;
        private final String videoPath;
        private final RandomAccessFile videoFile;
        private final long fileSize;
        private final BlockingQueue<RequestFrame> requestsQueue;
        private final ReentrantLock lock = new ReentrantLock();
        private final ReentrantLock pauseLock = new ReentrantLock();
        private final CountDownLatch killEvent = new CountDownLatch(1);
        private final Thread workerThread;

        private WorkerStreamLayer(UDPChannel udpChannel, String videoPath) throws IOException {
            this.udpChannel = udpChannel;
            this.videoPath = videoPath;
            this.videoFile = new RandomAccessFile(videoPath, "r");
            this.fileSize = videoFile.length();
            SessionSettings sessionSettings = new SessionSettings(); // Create an instance of SessionSettings
            this.requestsQueue = new LinkedBlockingQueue<>(sessionSettings.getMaxRequests());
            this.workerThread = new Thread(this::handlerVideo);
            this.workerThread.start();
        }

        private void processRequest(RequestFrame requestFrame) {
            lock.lock();
            try {
                byte[] data = new byte[requestFrame.getSize()];
                int bytesRead = videoFile.read(data);
                if (bytesRead > 0) {
                    udpChannel.sendData(data);
                }
            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                lock.unlock();
            }
        }

        private void handlerVideo() {
            while (true) {
                try {
                    pauseLock.lock();
                    try {
                        if (killEvent.getCount() == 0) {
                            return;
                        }

                        RequestFrame requestFrame = requestsQueue.poll(1, TimeUnit.SECONDS);
                        if (requestFrame != null) {
                            processRequest(requestFrame);
                        }
                    } finally {
                        pauseLock.unlock();
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        }

        public void stop() throws IOException {
            lock.lock();
            try {
                videoFile.close();
            } finally {
                lock.unlock();
            }
            killEvent.countDown();
            try {
                workerThread.join();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            udpChannel.close();
        }
    }

    private final WorkerStreamLayer workerStreamLayer;

    public StreamLayer(UDPChannel udpChannel, String videoPath) throws IOException {
        this.workerStreamLayer = new WorkerStreamLayer(udpChannel, videoPath);
    }

    public void addRequest(RequestFrame requestFrame) {
        try {
            workerStreamLayer.requestsQueue.put(requestFrame);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    public boolean updateOffset(long offset) {
        if (offset >= workerStreamLayer.fileSize) {
            return false;
        }
        workerStreamLayer.lock.lock();
        try {
            workerStreamLayer.videoFile.seek(offset);
            return true;
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        } finally {
            workerStreamLayer.lock.unlock();
        }
    }

    public void pause() {
        workerStreamLayer.pauseLock.lock();
    }

    public void stop() throws IOException {
        workerStreamLayer.stop();
    }

    @SuppressWarnings("removal")
    @Override
    protected void finalize() throws Throwable {
        try {
            stop();
        } finally {
            super.finalize();
        }
    }
}